File: Problems/987_Vertical_Order_Traversal_of_a_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    map<int, vector<pair<int, int>>> hm;
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        dfs(root, 0, 0);
        vector<vector<int>> res;
        for (auto it : hm) {
            sort(it.second.begin(), it.second.end());
            vector<int> tmp;
            for (auto i : it.second) {
                tmp.push_back(i.second);
            }
            res.push_back(tmp);
        }
        return res;
    }
    void dfs(TreeNode* node, int level, int order) {
        if (! node) return;
        hm[order].push_back({level, node -> val});
        dfs(node -> left,  level + 1, order - 1);
        dfs(node -> right, level + 1, order + 1);
    }
};

File: Problems/987_Vertical_Order_Traversal_of_a_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        hm = defaultdict(list)
        def dfs(node, order, level):
            if not node:
                return
            hm[order].append((level, node.val))
            dfs(node.left,  order - 1, level + 1)
            dfs(node.right, order + 1, level + 1)
        dfs(root, 0, 0)
        res = []
        for i in sorted(hm.keys()):
            tmp = []
            for j in sorted(hm[i]):
                tmp.append(j[1])
            res.append(tmp)
        return res